### PR TITLE
Fix dropping sharding on 0-sized boolean masks (#35273)

### LINKED ISSUE #35273: Inconsistent value dependent device for boolean masked array
### Description

On a colab notebook with a CUDA device I observe the following:

```python
import numpy as np
import jax.numpy as jnp
import jax

cpu_device = jax.devices("cpu")[0]
cpu_array = jnp.asarray([1], device=cpu_device)
masked = cpu_array[jnp.asarray([True], device=cpu_device)]
masked.shape, masked.device
```
```
((1,), CpuDevice(id=0))
```

Which is what I would expect.

However, when the masking result in a 0-sized dimension, I get a result on the CUDA device instead of the CPU device I would have expected given the explicit device selection the above code:

```python
masked = cpu_array[jnp.asarray([False], device=cpu_device)]
masked.shape, masked.device
```
```
((0,), CudaDevice(id=0))
```

### System info (python version, jaxlib version, accelerator, etc.)

```


jax:    0.7.2
jaxlib: 0.7.2
numpy:  2.0.2
python: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]
device info: Tesla T4-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='966a76ee8f69', release='6.6.105+', version='#1 SMP Thu Oct  2 10:42:05 UTC 2025', machine='x86_64')

$ nvidia-smi
Fri Feb 20 16:12:03 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   71C    P0             32W /   70W |   11305MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            2061      C   /usr/bin/python3                      11302MiB |
+-----------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by Ashutosh0x**
Fixes #35273.

This fixes an issue where boolean masking resulting in a 0-sized dimension would drop the array's sharding and device placement, falling back to the default device (e.g., CUDA instead of CPU).

The root cause was in `jax.lax.gather`'s empty-slice fast-path (`jax/_src/numpy/indexing.py`). It deliberately delegates to `jax.lax.full_like` and passes `indexer.slice_sharding`, which is derived from the array's abstract `ShapedArray` sharding. For single-device arrays, this is equivalent to `NamedSharding(empty_mesh)`. By explicitly passing an empty mesh sharding to `full_like`, it overrides the normal fallback to the original array's physical `SingleDeviceSharding`.

This fix drops the abstract `slice_sharding` if its mesh is empty before calling `full_like`, ensuring it falls back correctly to the inheritance of the original array's physical device.

### CODE REVIEW

**Comment by yashk2810** [comment]
Hi!

Thank you for the PR but we can fix it. I already rejected your first attempt and this is also not correct.

I appreciate the eagerness to fix, but this fix requires more context and sadly AI won't be able to help here :)

So i am going to close this PR too.

**Comment by Ashutosh0x** [comment]
Hi @yashk2810,

Thank you for reviewing the PR and taking the time to provide feedback! I completely understand if this PR isn't the correct architectural approach for JAX. 

I was attempting to trace why the device placement gets dropped during boolean masking that results in a 0-sized dimension. From my tracing, the fallback to the default device seems to happen because `jax._src.numpy.indexing._gather` hits an empty-slice fast-path and calls `lax.full_like`. 

During this call, it explicitly passes `indexer.slice_sharding`, which for single-device arrays evaluates to an abstract `NamedSharding(empty_mesh)` via `core.typeof(arr).sharding`. Because `full_like` is given this explicit empty-mesh sharding, it overrides any fallback to the original array's physical `SingleDeviceSharding`, ultimately placing the 0-sized array on the default value-dependent device (e.g. CUDA instead of CPU).

My intent with this PR was to see if dropping that abstract empty-mesh `slice_sharding` would allow `full_like` to naturally inherit the physical device again (which appeared to work in simple local tests). 

However, since you mentioned this fix is incorrect and requires more context, I'd genuinely love to learn why! Does `full_like` need to be modified instead to handle the abstraction correctly, or is the dropping of device placement for 0-sized masked arrays intended behavior for a different reason entirely? 

Any pointers or context you can share would be incredibly helpful. Thank you again!

**Comment by Ashutosh0x** [comment]
Hi @yashk2810! Thanks again for the feedback.

I took a deeper look into why the sharding was being dropped and realized my previous fix in `indexing.py` was essentially a band-aid. After looking at your commit `3b6fd04e0`, I saw that you've already implemented smart sharding propagation in `lax.full_like`, but it gets overridden if an explicit sharding is passed. 

In the 0-sized masking case, `indexing.py` passes down an abstract `NamedSharding(empty_mesh)` from the `aval`. Because this is not `None`, `full_like` uses it literally and loses the physical device info.

I've refactored the fix to be more architectural: I'm now modifying `lax.full_like` to treat abstract empty-mesh shardings as "unspecified". This allows your core propagation logic to correctly fall back to the physical sharding of the source array, preserving the device affinity (e.g. CPU) without high-level hacks in the indexing logic.

I've pushed this refined version to the branch. Would love to know if this aligns better with the JAX sharding model!

**Comment by Ashutosh0x** [comment]
Thanks @gemini-code-assist! I reviewed your feedback concerning the check for abstract meshes missing Tracer objects, and the invalid _is_concrete check on the mesh rather than the NamedSharding object. 

I've fixed these implementation bugs and pushed the correct architecture to #35296
